feat(content-manager): Select fields that require validation in availablelocales

jhoward1994 opened the pull request:
<!--
Hello 👋 Thank you for submitting a pull request.

To help us merge your PR, make sure to follow the instructions below:

- Create or update the tests
- Create or update the documentation at https://github.com/strapi/documentation
- Refer to the issue you are closing in the PR description: Fix #issue
- Specify if the PR is ready to be merged or work in progress (by opening a draft PR)

Please ensure you read the Contributing Guide: https://github.com/strapi/strapi/blob/main/CONTRIBUTING.md
-->

### What does it do?

Select fields that will require validation in available locales document metadata

### Why is it needed?

For displaying the validation errors in the multiple locale modal we need each available locale to at least contain all the fields that will require validation

### How to test it?

WIP

### Related issue(s)/PR(s)

DX-1364


#### Notes

- E2E EE History also failing on main

jhoward1994 commented:
@Marc-Roig this is wip but it looks like I will need something like this for the FE of the bulk locale action. what do you think of this?

jhoward1994 commented:
Hey @innerdvations 

Could you review this if you have time?

Looks like the E2E failures atm are caused by what we're working on in https://github.com/strapi/strapi/pull/20185

innerdvations commented on packages/core/content-manager/server/src/services/document-metadata.ts:
```diff
@@ -186,10 +234,11 @@ export default ({ strapi }: { strapi: Core.Strapi }) => ({
-      select: ['createdAt', 'updatedAt', 'locale', 'publishedAt', 'documentId'],
       populate: {
+        ...getDeepPopulate(uid),
         createdBy: {
```
(assuming this is true and I'm not wrong about how it's working) it might be worth a comment that without these manual createdBy and updatedBy, the deep populate will cause all creator fields to be present, otherwise they could end up removed and introduce a security issue. Or maybe add a test that ensures certain fields like password, reset token, etc, aren't present in the results so nobody is even able to break it.

innerdvations reviewed (commented):
LGTM overall, just a couple comments

I didn't QA since it's going into a feature branch, we can test it there

jhoward1994 commented:
Thanks @innerdvations 

I've added a test for the sensitive user data in [`e1933cb` (#20156)](https://github.com/strapi/strapi/pull/20156/commits/e1933cbb1d1176b33aaec8bf3f92fb1db23f7c38)

Theres something flaky going on though, the 2 tests I've added to that file can't be ran the other way round nor can they be ran in a transaction like the others in that file. I can't work out whats going on there. Do you have any ideas?

Marc-Roig commented on packages/core/content-manager/server/src/services/document-metadata.ts:
```diff
@@ -85,24 +94,63 @@ export default ({ strapi }: { strapi: Core.Strapi }) => ({
+          localeVersions,
+          async (localeVersion: DocumentVersion) =>
+            traverseEntity(
+              ({ key, attribute }, { remove }) => {
```
will that be the case when the entity has a key which does not exist on the schema?

Marc-Roig commented on packages/core/content-manager/server/src/services/document-metadata.ts:
```diff
@@ -233,8 +290,43 @@ export default ({ strapi }: { strapi: Core.Strapi }) => ({
     const meta = await this.getMetadata(uid, document, opts);
+    let { availableLocales, availableStatus } = meta;
+
+    // Always sanitize passwords
+    const defaultSanitizer = sanitize.sanitizers.sanitizePasswords({
+      schema: strapi.getModel(uid),
+      getModel(uid: string) {
+        return strapi.getModel(uid as UID.Schema);
+      },
+    });
```
could we create the sanitizer only once, so there is no need to duplicated the availableLocales & status code?

Also, shouldn't it be more adecuate to sanitize data on the controller level? even if it's duplicated, we do the same for the document-manager service.

Good catch on sanitizing those fields!

Marc-Roig replied:
Which leds me to think, what happens if a user doesn't have read permission of a field required to check the publishing validations?

innerdvations reviewed (approved):
Nice! I didn't QA because I don't know exactly where and how it will be used, it but code LGTM now! Let me know if you do need me for that though

innerdvations commented:
Oops, looks like there's a build error from types

jhoward1994 commented on packages/core/core/src/services/document-service/entries.ts:
```diff
@@ -15,6 +15,7 @@ const createEntriesService = (uid: UID.ContentType) => {
   const contentType = strapi.contentType(uid);
 
   async function createEntry(params = {} as any) {
+    // TODO fails here
```
Oh dear I dont think I meant to commit this, it was a local reminder will clean this up

jhoward1994 commented on tests/api/core/content-manager/api/document-metadata.test.api.js:
```diff
@@ -11,12 +11,54 @@ let formatDocument;
+    },
+    // TODO - Add other fields with validation requirements
+    shopName: {
+      // This field has no validation requirements so should be expluded from
```
Agreed, sounds like a really anti climactic explosion

jhoward1994 commented:
I don't get a build error locally, it looks like that bit of code should pass fine to me. Hopefully this CI run passes

innerdvations reviewed (approved):
LGTM!
